test/e2e/json-rpc/wallet_addEthereumChain.spec.ts

```diff
@@ -0,0 +1,329 @@
+              {
+                chainId: '0x53a',
+                chainName: 'Localhost 1338',
+                nativeCurrency: {
+                  name: '',
+                  symbol: 'ETH',
+                  decimals: 18,
+                },
+                rpcUrls: ['http://localhost:8546'],
+                blockExplorerUrls: [],
+              },
+            ],
+          });
+
+          await driver.executeScript(
+            `window.ethereum.request(${switchEthereumChainRequest})`,
+          );
+
+          await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+          await driver.clickElement({ text: 'Approve', tag: 'button' });
+
+          await driver.switchToWindowWithTitle(WINDOW_TITLES.TestDApp);
+
+          const afterPermittedChains = await getPermittedChains(driver);
+          assert.deepEqual(afterPermittedChains, ['0x53a']);
+        },
+      );
+    });
+
+    it('automatically permits and switches to the chain when the rpc endpoint is added but a different rpc endpoint already existed for the chain', async function () {
+      await withFixtures(
+        {
+          dapp: true,
+          fixtures: new FixtureBuilder()
+            .withNetworkControllerDoubleGanache()
+            .build(),
+          ganacheOptions: {
+            ...defaultGanacheOptions,
+            concurrent: [{ port: 8546, chainId: 1338 }],
+          },
+          title: this.test?.fullTitle(),
+        },
+        async ({ driver }: { driver: Driver }) => {
+          await unlockWallet(driver);
+          await openDapp(driver);
+
+          const beforePermittedChains = await getPermittedChains(driver);
+
+          assert.deepEqual(beforePermittedChains, []);
+
+          const switchEthereumChainRequest = JSON.stringify({
+            jsonrpc: '2.0',
+            method: 'wallet_addEthereumChain',
+            params: [
+              {
+                chainId: '0x53a',
+                chainName: 'Localhost 8546 alternative',
+                nativeCurrency: {
+                  name: '',
+                  symbol: 'ETH',
+                  decimals: 18,
+                },
+                // this does not match what already exists in the NetworkController
+                rpcUrls: ['http://127.0.0.1:8546'],
+                blockExplorerUrls: [],
+              },
+            ],
+          });
+
+          await driver.executeScript(
+            `window.ethereum.request(${switchEthereumChainRequest})`,
+          );
+
+          await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+          await driver.clickElement({ text: 'Approve', tag: 'button' });
+
+          await driver.switchToWindowWithTitle(WINDOW_TITLES.TestDApp);
+
+          const afterPermittedChains = await getPermittedChains(driver);
+          assert.deepEqual(afterPermittedChains, ['0x53a']);
+        },
+      );
+    });
+
+    it('prompts to switch to the chain when the rpc endpoint being added already exists', async function () {
```

Comment:
are there two different cases between when the chain is currently selected and when its not?

Reply:
yeah i suppose so. if the network being added already exists and the dapp is already on the chainId, nothing happens, even if the dapp isn't currently permitted to use the chain yet. That's the current behavior in main though and I'm trying not to bite off too much with each PR. I'm unsure if that behavior is right, but it's what's in main and not what i'm targeting here